src/paint/map_element/scenery.c

```diff
@@ -16,8 +16,309 @@
+				} else {
+					sub_98199C(image_id, x_offset, y_offset, boxlength.x, boxlength.y, boxlength.z - 1, height, boxoffset.x, boxoffset.y, boxoffset.z, get_current_rotation());
+				}
+			}
+		}
+	}
+	// 6E0556: Draw supports:
+	if (mapElement->properties.scenery.colour_1 & 0x20) {
+		if (!(entry->small_scenery.flags & SMALL_SCENERY_FLAG20)) {
+			int ax = 0;
+			int supportHeight = height;
+			if (supportHeight & 0xF) {
+				supportHeight &= 0xFFFFFFF0;
+				ax = 49;
+			}
+			uint32 supportImageColourFlags = 0x20000000;
+			if (entry->small_scenery.flags & SMALL_SCENERY_FLAG26) {
+				supportImageColourFlags = ((mapElement->properties.scenery.colour_1 & 0x1F) << 19) | 0x20000000;
+			}
+			if (dword_F64EB0 != 0) {
+				supportImageColourFlags = dword_F64EB0;
+			}
+			if (direction & 1) {
+				wooden_b_supports_paint_setup(1, ax, supportHeight, supportImageColourFlags);
+			} else {
+				wooden_b_supports_paint_setup(0, ax, supportHeight, supportImageColourFlags);
+			}
+		}
+	}
+	// 6E05D1:
+	height += entry->small_scenery.height;
+	uint16 word_F64F2A = height;
+	height += 7;
+	height &= 0xFFF8;
+	paint_util_set_general_support_height(height, 0x20);
+	// 6E05FF:
+	if (entry->small_scenery.flags & SMALL_SCENERY_FLAG23) {
+		height = word_F64F2A;
+		if (entry->small_scenery.flags & SMALL_SCENERY_FLAG_FULL_TILE) {
+			// 6E0825:
+			paint_util_set_segment_support_height(SEGMENT_C4, height, 0x20);
+			if (entry->small_scenery.flags & SMALL_SCENERY_FLAG_VOFFSET_CENTRE) {
+				paint_util_set_segment_support_height(SEGMENTS_ALL & ~SEGMENT_C4, height, 0x20);
+			}
+			return;
+		}
+		if (!(entry->small_scenery.flags & SMALL_SCENERY_FLAG_VOFFSET_CENTRE)) {
+			return;
+		}
+		int cl = ((mapElement->type >> 6) + get_current_rotation()) & 3;
+		// 6E075C:
+		switch (cl) {
+		case 0:
+			paint_util_set_segment_support_height(SEGMENT_B4 | SEGMENT_C8 | SEGMENT_CC, height, 0x20);
+			break;
+		case 1:
+			paint_util_set_segment_support_height(SEGMENT_BC | SEGMENT_CC | SEGMENT_D4, height, 0x20);
+			break;
+		case 2:
+			paint_util_set_segment_support_height(SEGMENT_C0 | SEGMENT_D0 | SEGMENT_D4, height, 0x20);
+			break;
+		case 3:
+			paint_util_set_segment_support_height(SEGMENT_B8 | SEGMENT_D0 | SEGMENT_C8, height, 0x20);
+			break;
+		}
+		return;
+	}
+	if (entry->small_scenery.flags & (SMALL_SCENERY_FLAG27 | SMALL_SCENERY_FLAG_FULL_TILE)) {
+		paint_util_set_segment_support_height(SEGMENT_C4, 0xFFFF, 0);
+		if (entry->small_scenery.flags & SMALL_SCENERY_FLAG_VOFFSET_CENTRE) {
+			paint_util_set_segment_support_height(SEGMENTS_ALL & ~SEGMENT_C4, 0xFFFF, 0);
+		}
+		return;
+	}
+	if (!(entry->small_scenery.flags & SMALL_SCENERY_FLAG_VOFFSET_CENTRE)) {
+		return;
+	}
+	int cl = ((mapElement->type >> 6) + get_current_rotation()) & 3;
+	switch (cl) {
+	case 0:
+		paint_util_set_segment_support_height(SEGMENT_B4 | SEGMENT_C8 | SEGMENT_CC, 0xFFFF, 0);
```

Comment:
By using `paint_util_rotate_segments(SEGMENT_B4 | SEGMENT_C8 | SEGMENT_CC, direction)`, you can remove these switches.

Here's a reminder of the grid (top row would be the top right edge of a tile)

<pre>
<b>B4 CC</b> BC
<b>C8</b> D4 D4
B8 D0 C0
</pre>